### PR TITLE
Enabling `imageGPT` auto feature extractor.

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Attempts to superseed https://github.com/huggingface/transformers/pull/16869
With less specific overrides Thanks @ydshieh, you definitely couldn't guess about the options I modified here :)

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/pipelines/base.py [hunk 1]
```diff
@@ -75,7 +75,7 @@ def _pad(items, key, padding_value, padding_side):
         # Others include `attention_mask` etc...
         shape = items[0][key].shape
         dim = len(shape)
-        if dim == 4:
+        if key == "pixel_values":
```
Comment (Narsil) [inline src/transformers/pipelines/base.py:78]: `pixel_values` is a consistent key used across `transformers` that doesn't require padding so I think it's ok to update that condition to cover ImagePGT

File: tests/imagegpt/test_modeling_imagegpt.py [hunk 2]
```diff
@@ -171,6 +171,12 @@ def get_config(
             reorder_and_upcast_attn=reorder_and_upcast_attn,
         )
 
+    def get_pipeline_config(self):
+        config = self.get_config()
+        config.vocab_size = 513
```
Comment (Narsil) [inline tests/imagegpt/test_modeling_imagegpt.py:176]: This is necessary because the feature extractor will not respect `vocab_size` (it takes actual pixel values) so we need to use a large enough vocab and sequence to fit real world images used in those tests.

This modified only the pipeline tests so we can keep the original model tests strictly similar.

File: src/transformers/pipelines/base.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/pipelines/base.py:78]: How does that solve the problem since currently the [feature extractor of ImageGPT is outputting pixel values](https://github.com/huggingface/transformers/blob/daf520b03345e270fa8cca81c4dbc3d9f46b413c/src/transformers/models/imagegpt/feature_extraction_imagegpt.py#L174)?

**Review by sgugger — APPROVED** [review]
LGTM even if I don't understand why it works.

**Comment by ydshieh** [comment]
@sgugger For `ImageGPT`, we don't want to perform padding (otherwise, it fails with `None` as padding value).

As you mentioned,  `ImageGPTFeatureExactor` outputs `pixel_values` (despite it is not a very good name). So @Narsil 's change

https://github.com/huggingface/transformers/blob/104ee5b3fe1b164d71dec6efd8ddd00a381d13ba/src/transformers/pipelines/base.py#L78-L81

will skip padding.

File: src/transformers/pipelines/base.py (hunk 1 above)
Comment (Narsil) [reply src/transformers/pipelines/base.py:78]: The key part is this : `        # flatten to (batch_size, height*width)`

So the tensor is `dim == 2` even though it does represent `pixel_values`. 
Since the `dim==4` was intended to catch images and afaik there are no other image like tensor aside from `pixel_values` I am switching condition to catch the `dim==2` tensor too.

Images don't get padding basically, the feature extractor is responsible for making them of the same size so they can be stacked

**Comment by Narsil** [comment]
> 
> will skip padding. (Although the comment could be more specific regarding the exceptional case)

 I added this `# ImageGPT actually will use B, SEQ_LEN not tensor of shape 4` to specify there's an edge case, do you think we could phrase this better ?

**Comment by ydshieh** [comment]
> > will skip padding. (Although the comment could be more specific regarding the exceptional case)
> 
> I added this `# ImageGPT actually will use B, SEQ_LEN not tensor of shape 4` to specify there's an edge case, do you think we could phrase this better ?

This is good enough for me 😄

File: src/transformers/pipelines/base.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/pipelines/base.py:78]: But ImageGPT using `pixel_values` as a key is a mistake fixed [here](https://github.com/huggingface/transformers/pull/16872) so this test won't catch `ImageGPT` anymore.

Comment (Narsil) [reply src/transformers/pipelines/base.py:78]: I see, then will the feature extractor provide a  `padding_value`?   Would it make sense in that context? 

If yes, then we don't need to modify anything here, it will pad nicely automatically.
If no, this means that padding (hence batching at inference) doesn't really mean anything and we shouldn't rely on it, no ? Or are the `input_ids` produces always of the same `seq_len` (in which case we can go for something like @ydshieh proposed which is to pad only when tensors stack even if  `padding_value` does not exist.)

Comment (sgugger) [reply src/transformers/pipelines/base.py:78]: I think the processed images will all be of the same size.

**Comment by sgugger** [comment]
Merging #16872 so you can rebase on it and make sure any fix works with that change.

**Comment by Narsil** [comment]
Hi @sgugger ,

Lost track of this one, is it ok to merge ?

**Comment by sgugger** [comment]
Good for me as long as the tests passing are on a rebased main. Maybe @NielsRogge can have a look too to confirm it's good to go?

**Review by NielsRogge — APPROVED** [review]
LGTM!

**Comment by nguyenhoanghai1605** [comment]
To enable the feature extractor for ImageGPT, use `AutoFeatureExtractor` from Hugging Face:

```python
from transformers import AutoFeatureExtractor

extractor = AutoFeatureExtractor.from_pretrained("openai/imagegpt-small")
